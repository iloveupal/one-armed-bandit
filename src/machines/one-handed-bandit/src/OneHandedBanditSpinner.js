import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';

import SlotMachine from './components/SlotMachine';
import SlotComponent from './components/SlotComponent';

import {
    selectRandomSlotValues,
    initializeSpinnerIntervals,
    getNextSlotsState,
    generateSlotGraph,
} from './OneHandedBanditUtils';


export default class OneHandedBanditSpinner extends PureComponent {
    static propTypes = {
        isSpinning: PropTypes.bool.isRequired,
        slotsCount: PropTypes.number.isRequired,
        possibleSlotValues: PropTypes.arrayOf(PropTypes.string.isRequired).isRequired,
        slotImagesMap: PropTypes.object.isRequired,
        onRoundFinished: PropTypes.func.isRequired,
    }

    spinnerIntervals = [];
    // a map that will help the update fn transition between possible slot values faster.
    // will look like { <prevSlotValue>: <nextSlotValue> }; 
    slotGraph = null;

    constructor (props) {
        super(props);

        this.slotGraph = generateSlotGraph(props.possibleSlotValues);

        this.state = {
            slots: selectRandomSlotValues(props.slotsCount, props.possibleSlotValues),
        };
    }

    componentDidUpdate(prevProps) {
        if ( prevProps.isSpinning === false && this.props.isSpinning === true ) {
            this.startSpinning();
        }

        if ( prevProps.isSpinning === true && this.props.isSpinning === false ) {
            this.stopSpinning();
        }
    }

    startSpinning = () => {
        this.spinnerIntervals = initializeSpinnerIntervals(this.props.slotsCount, this.updateSlot);
    }

    stopSpinning = () => {
        this.spinnerIntervals.forEach((timerId) => clearInterval(timerId));
        this.spinnerIntervals = [];

        this.props.onRoundFinished(this.state.slots);
    }

    updateSlot = (slotIndex) => {
        this.setState((prevState) => {
            return {
                slots: getNextSlotsState({
                    currentSlotsState: prevState.slots,
                    slotGraph: this.slotGraph,
                    slotIndexToUpdate: slotIndex,
                }),
            };
        });
    }

    render () {
        return (
            <SlotMachine>
                { this.state.slots.map((slot, index) => (
                    <SlotComponent
                        slot={slot}
                        slotImagesMap={this.props.slotImagesMap}
                        key={index}
                    />
                )) }
            </SlotMachine>
        );
    }
}
