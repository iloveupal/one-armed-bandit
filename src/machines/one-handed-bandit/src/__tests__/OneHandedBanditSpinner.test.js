import React from 'react';
import renderer from 'react-test-renderer';
import 'jest-styled-components'

import OneHandedBanditSpinner from '../OneHandedBanditSpinner';


describe('OneHandedBanditSpinner Component', () => {
    it('should match the snapshot', () => {
        const component = renderer.create(
            <OneHandedBanditSpinner
                isSpinning={false}
                slotsCount={4}
                possibleSlotValues={['banana']}
                slotImagesMap={{
                    banana: 'img'
                }}
                onRoundFinished={jest.fn()}
            />
        );

        const tree = component.toJSON();

        expect(tree).toMatchSnapshot();
    });

    it('should call internal method startSpinning when the prop isSpinning changes to true', () => {
        const startSpinning = jest.fn();

        const component = renderer.create(
            <OneHandedBanditSpinner
                isSpinning={false}
                slotsCount={3}
                possibleSlotValues={['banana']}
                slotImagesMap={{
                    banana: 'img'
                }}
                onRoundFinished={jest.fn()}
            />
        );

        component.root.instance.startSpinning = startSpinning;

        component.update(
            <OneHandedBanditSpinner
                isSpinning={true}
                slotsCount={3}
                possibleSlotValues={['banana']}
                slotImagesMap={{
                    banana: 'img'
                }}
                onRoundFinished={jest.fn()}
            />
        );

        expect(startSpinning).toHaveBeenCalledTimes(1);
    });

    
    it('should call the onRoundFinished prop when the prop isSpinning changes to false', () => {
        const onRoundFinished = jest.fn();
        
        const component = renderer.create(
            <OneHandedBanditSpinner
                isSpinning={true}
                slotsCount={3}
                possibleSlotValues={['banana']}
                slotImagesMap={{
                    banana: 'img'
                }}
                onRoundFinished={onRoundFinished}
            />
        );

        component.update(
            <OneHandedBanditSpinner
                isSpinning={false}
                slotsCount={3}
                possibleSlotValues={['banana']}
                slotImagesMap={{
                    banana: 'img'
                }}
                onRoundFinished={onRoundFinished}
            />
        );

        expect(onRoundFinished).toHaveBeenCalledWith(['banana', 'banana', 'banana']);
    });
});
