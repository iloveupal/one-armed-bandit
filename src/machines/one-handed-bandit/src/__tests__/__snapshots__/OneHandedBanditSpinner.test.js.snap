// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OneHandedBanditSpinner Component should match the snapshot 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-flow: row no-wrap;
  -ms-flex-flow: row no-wrap;
  flex-flow: row no-wrap;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  padding: 20px 0;
}

.c1 {
  width: 100px;
  height: 100px;
}

<div
  className="c0"
>
  <img
    className="c1"
    src="img"
  />
  <img
    className="c1"
    src="img"
  />
  <img
    className="c1"
    src="img"
  />
  <img
    className="c1"
    src="img"
  />
</div>
`;
